### DISCUSSION TITLE
When should we be using multiple stores instead of a single store with separate slices?

### DISCUSSION THREAD

**Discussion by lego-sharat** [pmndrs/zustand#2496] [Q&A] [answered]
Does it depend on size of the store? Is there a rule of thumb that should be followed?

**Comment by dbritto-dev** [comment]
We follow just the AHA programming and is basically when you feel that something is getting difficult to maintain, that's the moment to start splitting your store or using slices. How can I know if I need to use multiple stores or multiple slices? well that's simple, if each part is independent from each other then multiple stores is the right path if not then multiple slices is the right path

**Comment by dai-shi** [comment, marked as answer]
My suggestion (as of now):
- If two things are totally isolated, multiple stores would be good.
- If they are slightly related, a single store would be good, because in the future you want an action to update both.
- For the context use case, a single store might be easier, because you need only one React Context.

**Comment by Sebastian-Nielsen** [comment]
@dai-shi @dbritto-dev 
You are answers are conflicting with what's writting in the docs. 

You both advocate for a single store in some situations and multiple stores in others in this discussion thread. However, on the docs page it says a single zustand store is best practice: ["Although Zustand is an unopinionated library, we do recommend a few patterns. [...] Your applications global state should be located in a single Zustand store."  
](https://docs.pmnd.rs/zustand/guides/flux-inspired-practice)

@dai-shi you say "If two things are totally isolated, multiple stores would be good". Okay, but would it be better than a single store with two slices, where each isolated part is stored in its own slice? If not, well then you are actually advocating for a single store. 

It's confusing that the maintainer of zustand recommends something that conflicts with the docs. Is the docs just outdated or what?

**Reply by dai-shi** [reply]
Now, as we will document context usage, I think the single store would be the default and multiple stores are possible.

(Nothing is conflicting. The baseline is it's unopinionated. But, I understand it's confusing.)

**Reply by dbritto-dev** [reply]
@Sebastian-Nielsen sorry about that, we should update the docs and make clear that you shouldn't always use one global store for everything. Actually, there're multiple patterns. 

By the way, I'm working on two things: update docs content and update website that will use vitepress in the new version.

**Comment by sanjeevyadavIT** [comment]
Can someone explain what pattern should I use for an ecommerce app.

for exmaple, I have Screens 
1. Profile Screen that will have user related info
2. Cart Screen that will have items in cart
3. Product Detail Screen 

Should I use slice pattern and create 1 global store where user data, cart data and product detail data reside

or should I create three different store like `useUserStore`, `useCartStore`, `useProductStore` ?

**Reply by dbritto-dev** [reply]
Tip:
- for dependent stores is better one big store 
- for independent stores is better multiple stores

**Reply by dbritto-dev** [reply]
@lego-sharat https://github.com/pmndrs/zustand/discussions/2496#discussioncomment-9176980

**Reply by deanolium** [reply]
For this, I would probably go with three stores, since the user info isn't going to be used or updated at the same time as the cart or as the products. 

You could argue this down into two stores, depending on how you do the cart and product data, and how you use them. For instance, the cart could just hold productIDs, and then when you display what's in the basket, you have a hook which pulls the product details and the cart contents out, then map the IDs to the details and return that - so then the end consumer just calls this derived data hook and gets the full list of product data for everything in the basket, whilst still having a single source of truth.

Maybe slice the basket from the product store, which then allows you to keep the code separate and let you more easily test them independently, whilst still having them in a single store. Though you could have them as separate stores and still use your derived data hook which pulls in from both stores and then combines them.

There's lots of options here, but some of it is down to how complex the stores will be and your own taste. To be honest, in the real world you'll be more likely to grab the product details from a server and they won't be changed in the app. So using TanStack Query to hold that data would make more sense. The only real use-case for storing product details in a global store is if you're doing an offline app which might need to mutate the product details for some reason. I mainly work on online experiments for clients where this is a use-case for us, but it won't be the case for 99% of people.

**Reply by mooalot** [reply]
I would go with a single store. There is nothing more frustrating than using data from multiple stores to get one result.
